src/AsyncMethodTester.mo

```diff
@@ -20,76 +16,57 @@ module {
-    before : TReq -> TState,
-    after : TState -> TResp,
-    iterations_limit : ?Nat,
-  ) {
-    type Request<T> = {
-      #error;
-      #arg : T;
-    };
-
-    func requsetToData(r : Request<TReq>) : Data<TReq, TState, TResp> {
-      let #arg x = r else return #error;
-      #before x;
-    };
-
-    type Data<TReq, TState, TResp> = {
-      #error;
-      #before : TReq;
-      #middle : TState;
-      #after : TResp;
-    };
-
-    type Call<TReq, TState, TResp> = {
+  public class AsyncMethodTester<T, S, R>(iterations_limit : ?Nat) {
+    type Response<T, S, R> = {
       var lock : Bool;
       var state : State;
-      var data : ?Data<TReq, TState, TResp>;
-    };
-
-    func new<TReq, TResp>() : Call<TReq, TState, TResp> = {
-      var lock = true;
-      var state = #empty;
-      var data = null;
+      methods : ?(T -> S, S -> R);
     };
 
     let limit = Option.get(iterations_limit, 100);
-    var calls : Buffer.Buffer<Call<TReq, TState, TResp>> = Buffer.Buffer<Call<TReq, TState, TResp>>(2);
-    var resp_count = 0;
-    var req_count = 0;
-    var last_call_result : ?TResp = null;
-
-    public func stage(arg : Request<TReq>) : Nat {
-      if (calls.size() == resp_count) {
-        calls.add(new());
+    var queue : Deque.Deque<Response<T, S, R>> = Deque.empty<Response<T, S, R>>();
+    var last_call_result : ?R = null;
+
+    public func stage(arg : ?(T -> S, S -> R)) : ReleaseState {
+      let response : Response<T, S, R> = {
+        var lock = true;
+        var state = #staged;
+        methods = arg;
       };
 
-      let call = calls.get(resp_count);
-      if (not Option.isNull(call.data)) {
-        Debug.trap("Data must be null before stage");
-      };
-      call.data := ?requsetToData(arg);
+      queue := Deque.pushBack(queue, response);
 
-      let ret = resp_count;
-      resp_count += 1;
+      object {
+        public func release() {
+          if (not response.lock) {
+            Debug.trap("Response must be locked before release");
+          };
+          response.lock := false;
+        };
 
-      ret;
+        public func state() : State = response.state;
+      };
     };
 
-    public func release(index : Nat) {
-      let call = calls.get(index);
-      if (not call.lock) {
-        Debug.trap("Call must be locked before release");
+    public func call(arg : T) : async* () {
+      var inc = limit;
+      while (Deque.isEmpty(queue)) {
+        await async ();
+        inc -= 1;
       };
-      call.lock := false;
-    };
 
-    public func state(index : Nat) : State = calls.get(index).state;
+      let ?(r, q) = Deque.popFront(queue) else Debug.trap("No response staged");
+      queue := q;
 
-    func await_unlock(r : Call<TReq, TState, TResp>) : async* () {
       r.state := #running;
-      var inc = limit;
+      let s = Option.apply<(T -> S, S -> R), S>(
+        r.methods,
+        ?(
+          func((pre, _)) {
+            pre(arg);
+          }
+        ),
+      );
```

Comment:
This part is probably easier readable with a switch statement instead of using Option.apply.